Class-dict synthesis algorithm refactor with errors and logging

danieldjohnson opened the pull request:
edit: This PR now maintains the backtracking behavior, while still adding better errors and logging (see https://github.com/google-research/dex-lang/pull/480#issuecomment-769554704)

---

**Generate logging outputs while running inference.**

Adds a new typeclass ExceptWithOutputs, which is just a combination of the
except and writer monads. This makes it possible to log outputs during type
inference, which can be useful for debugging.

**Redesign synthesis algorithm to make it more interpretable.**
The old synthesis algorithm essentially did a big exhaustive depth-first search
over all possible ways of building an instance, then finally raised an error if
there wasn't exactly one solution. This has a few non-ideal consequences:
- Errors always show up for the requested instance, instead of showing up for
  the child instance that we couldn't solve.
- There's no way to observe the sequence of synthesis decisions taken by the
  algorithm.
- Part of determining whether an instance matches involves solving for all of
  its own constraints. Thus if one of an instance's new constraints can't be
  solved, we might conclude that this instance doesn't apply, and continue
  searching for other instances instead of throwing an error. Effectively,
  we perform an unlimited amount of backtracking while solving instances.

This algorithm instead synthesizes constraints one at a time, without
recursively solving for all newly-introduced constraints immediately. It also
modifies the synthesis process to produce human-readable names for all of the
steps of synthesis, and writes them to the output when `%passes synth` is
active.

In the future, we may want to add a principled form of backtracking that keeps
the interpretability of this algorithm.

**Add solver context to error messages, and add tests for new errors.**

---

Issue #289 is an example of where the old synthesis algorithm's errors can lead people (including me) astray! After this PR, it looks like:
```
:t a
> ({left: Fin 3 | right: Fin 3} => Int32)

:p neg a
> Type error:Couldn't synthesize a class dictionary for: (VSpace Int32)
> While synthesizing (VSpace ({left: Fin 3 | right: Fin 3} => Int32)) using (instance20 : ((a:Type) ?-> (n:Type) ?-> (VSpace a) ?=> VSpace (n => a)))
>
> :p neg a
>    ^^^^
```
This makes it clear that the problem was that `Int32` isn't a `VSpace` under our definition (since you can't scale it by a float). Before, the error was just `Couldn't synthesize a class dictionary for: (VSpace ({left: Fin 3 | right: Fin 3} => Int32))`.

If you turn on `%passes synth` you can watch it solve things:
```
%passes synth
:p yieldAccum (AddMonoid Float) \ref.
     ref += [1.,2.,3.]
     ref += [2.,4.,5.]
> [3., 6., 8.]
> === synth ===
> Synthesizing ((h:Type) -> (AccumMonoid h Float32) ?=> AccumMonoid h ((Fin 3) => Float32)) using (tableAccumMonoid : ((h:Type)
>  ?-> (n:Type) ?-> (w:Type) ?-> (AccumMonoid h w) ?=> AccumMonoid h (n => w)))
> (after assuming premise:(AccumMonoid h Float32))
> (after assuming h:Type)
> === synth ===
> Synthesizing (AccumMonoid h Float32) using (tmp4 : (AccumMonoid h Float32))
> === synth ===
> Synthesizing (Add Float32) using (instance1 : (Add Float32))
> === synth ===
> Synthesizing (AccumMonoid a ((Fin 3) => Float32)) using (tableAccumMonoid : ((h:Type)
>  ?-> (n:Type) ?-> (w:Type) ?-> (AccumMonoid h w) ?=> AccumMonoid h (n => w)))
> === synth ===
> Synthesizing (AccumMonoid a Float32) using (tmp8 : (AccumMonoid a Float32))
> === synth ===
> Synthesizing (AccumMonoid a ((Fin 3) => Float32)) using (tableAccumMonoid : ((h:Type)
>  ?-> (n:Type) ?-> (w:Type) ?-> (AccumMonoid h w) ?=> AccumMonoid h (n => w)))
> === synth ===
> Synthesizing (AccumMonoid a Float32) using (tmp8 : (AccumMonoid a Float32))
> === synth ===
> Module (Core)
>   unevaluated decls:
     ...
>    _ans_:((Fin 3) => Float32) = tmp8
>
>   evaluated bindings:
>   ()
```

apaszke commented on src/lib/Builder.hs:
```diff
@@ -59,7 +59,8 @@ import PPrint
 
 newtype BuilderT m a = BuilderT (ReaderT BuilderEnvR (CatT BuilderEnvC m) a)
-  deriving (Functor, Applicative, Monad, MonadIO, MonadFail, Alternative)
+  deriving ( Functor, Applicative, Monad, MonadIO, MonadFail, MonadWriter w
```
What's `w` here?

danieldjohnson replied:
I believe it's implicitly a "forall `w`".

apaszke commented on tests/typeclass-tests.dx:
```diff
@@ -41,3 +41,81 @@ instance InterfaceTest4 Float
+  x
+> Type error:Multiple candidate class dictionaries for: (InterfaceSynthesisTest Float32)
+>   - (instance77 : (InterfaceSynthesisTest Float32))
+>   - (instance78 : (InterfaceSynthesisTest Float32))
```
IMO displaying auto-generated names is even worse than not displaying any names. We should really refine it to talk about source locations

danieldjohnson replied:
Yeah, this is uglier than I was expecting. I can play around with it to see if we can add source locations (but I'm not sure we expose enough information about filenames to type inference to make that work?)

danieldjohnson replied:
Seems like it's hard to pass source info through, so I'll just remove the names for now.

apaszke commented on src/lib/Inference.hs:
```diff
@@ -809,27 +824,41 @@ getSrcCtx = lift ask
+-- 1. Run type inference for every possible instance, and filter out those
+--    that do not unify. At this point, do not recursively synthesize any
+--    dictionaries.
+-- 2. Check that we got exactly one instance.
```
I don't think that this is an equivalently powerful algorithm to the old one. In particular, what if you had a hierarchy such as this one:
```
A a => B a
B a => C a
D a => E a
E a => C a
```
Now, consider a type that defines `E a`. In the first step of the search you cannot conclude failure just because you ended up with both `B a => C a` and `D a => E a` rules as applicable. Note that there is a single unique solution in this case and the previous algorithm would find it.

danieldjohnson replied:
Agreed, this is a less powerful algorithm. But I think that's the price you pay for better errors. Suppose a type `b` doesn't define anything, and then you request `C b`. What error do you show? If you're willing to do backtracking through everything, the best you can really do is say "I looked for `C b` but didn't find it". If you enforce that the synthesis algorithm only looks at one place in each iteration, you know exactly which missing instance was "responsible".

Haskell takes the non-backtracking approach, although I think some people have proposed various extensions that introduce backtracking in various principled ways.

This is related to #340, since in your example `B a => C a` and `E a => C a` can be seen as "overlapping" in the standard sense.

apaszke replied:
I really hate the overlapping instances restriction in Haskell. I feel like it's been pushing a lot of people either towards `{-# LANGUAGE OverlappingInstances #-}` which is a terrible hack to make the algorithm proceed even when there are multiple matches, or some arcane type-level programming to work around this limitation. I'd rather avoid replicating the same issue here. What if we printed out all partial derivations? I don't expect that there will be quite enough matches anyway

apaszke replied:
And even if there are more we can print something like:
```
Failed to derive the instance for <some expression>
The partial derivations are:
- <describe the derivation> (failed because an instance of <something> couldn't be found)
- <describe the derivation> (failed because an instance of <something> couldn't be found)
- <describe the derivation> (failed because an instance of <something> couldn't be found)
- ... (5 more derivations, specify `-some-flag` to print them all)
```

danieldjohnson replied:
Do you have real-world use cases similar to the scenario you mentioned above, where
- there are multiple possible strategies to derive `C a`,
- but only one of those strategies will succeed? (whereas overlapping instances in Haskell often is used to actually express a preference of one derivation over the other when multiple strategies succeed)

My personal feeling is that it's better for now to have clear error messages and an easily-interpretable algorithm, and then add more complexity back in once we actually have enough use cases. And my guess is that "unlimited backtracking but must find exactly one instance" isn't the right point to end up with, anyway; we probably want to handle overlapping instances even when multiple derivations succeed.

danieldjohnson replied:
After chatting with @dougalm, though, it seems like maybe it would be worthwhile to refactor the synthesis process a bit so that the algorithm/resolution strategy is orthogonal to the logging and inference machinery, since we may want to iterate on this in the future. I'll try to figure out something along those lines in the next week or so.

(Maybe a temporary compromise solution would be to run both the iterative and backtracking algorithms? If the iterative algorithm succeeds, use that; if it can't find anything, show a nice error; if it finds multiple things, then enable backtracking and see if that fixes the issue, but maybe show a worse error.)

danieldjohnson replied:
Looks like I was able to keep a fairly interpretable output while still doing backtracking. I'm still a bit skeptical that this is the "right" way to do it, but we can iterate on this more later if we have more concrete reasons to make a change.

apaszke commented on src/lib/Inference.hs:
```diff
@@ -840,51 +869,58 @@ traverseHoles fillHole = (traverseDecl recur, traverseExpr recur, synthPassAtom)
+      let m = do
+                ty' <- substBuilderR ty
+                inst <- step
+                instantiateAndCheck ty' inst
```
Can you please add a comment the describes what happens here?

danieldjohnson commented:
Ok, I've rewritten the algorithm again! I've re-introduced the backtracking behavior, so it should be equally powerful to the old algorithm. I've tried to keep the logging output interpretable:

- If synthesis fails, we show all of the dead-ends we reached, e.g. points where we were unable to make more progress. But we don't call out all of the other possible places a user could introduce a solution (for instance, by just declaring that the top-level instance exists).
- If we get multiple dictionaries, we show the first branching point where multiple decisions were produced. We don't show any other subsequent branching points that produce multiple decisions, partially because that's hard to understand, partially because it's hard to even implement.

This change also should hopefully make it easier to iterate on the synthesis algorithm in the future. I've refactored out all of the builder and inference logic and wrapped it in a Free monad, as an abstract computation that can ask for new subproblems and eventually produce a `Block`. The synthesis algorithm can then be a "pure" (non-monadic) recursive function that manipulates the results as needed.

axch commented:
This looks (from reading the discussion) like it might have been a nice contribution!  Unfortunately, at this point it has quite bit-rotted.  Is there any interest in porting (or rewriting) to the Safer Names system, and anything else that might have changed?  If not, let's close this PR.

danieldjohnson commented:
Yeah, this has definitely bit-rotted at this point. I'm not sure if/when I'll have time to rewrite this for Safer Names, and the changes will probably be big enough regardless that it would be better to start a new PR at that point.

Closing this now.
